WordPress/src/main/java/org/wordpress/android/ui/posts/editor/EditorMedia.kt

```diff
@@ -157,37 +150,37 @@ class EditorMedia(
     }
 
     fun advertiseImageOptimisationAndAddMedia(data: Intent) {
-        if (WPMediaUtils.shouldAdvertiseImageOptimization(activity)) {
-            WPMediaUtils.advertiseImageOptimization(activity) { addMediaItemGroupOrSingleItem(data) }
+        if (mediaUtilsWrapper.shouldAdvertiseImageOptimization()) {
+            mediaUtilsWrapper.advertiseImageOptimization { addMediaItemGroupOrSingleItem(data) }
         } else {
             addMediaItemGroupOrSingleItem(data)
         }
     }
 
     fun addExistingMediaToEditor(source: AddExistingMediaSource, mediaId: Long): Boolean {
-        val media = mediaStore.getSiteMediaWithId(site, mediaId)
-        if (media == null) {
-            AppLog.w(T.MEDIA, "Cannot add null media to post")
-            return false
+        mediaStore.getSiteMediaWithId(site, mediaId)?.let { media ->
+            editorTracker.trackAddMediaEvent(site, source, media)
+            fluxCUtilsWrapper.mediaFileFromMediaModel(media)?.let { mediaFile ->
+                editorMediaListener.appendMediaFile(mediaFile, media.urlToUse)
+            }
+            return true
         }
-
-        editorTracker.trackAddMediaEvent(site, source, media)
-
-        val mediaFile = FluxCUtils.mediaFileFromMediaModel(media)
-        editorMediaListener.appendMediaFile(mediaFile, media.urlToUse)
-        return true
+        AppLog.w(T.MEDIA, "Cannot add null media to post")
+        return false
     }
 
     fun addExistingMediaToEditor(source: AddExistingMediaSource, mediaIdList: List<Long>) {
         val mediaMap = ArrayMap<String, MediaFile>()
-        for (mediaId in mediaIdList) {
```

Comment:
This is just a discussion for fun, feel free to ignore it:D.

I usually prefer to avoid modifying variables from within a chain of operators (in this case we modify `mediaMap` from `forEach`).

So I'd probably write something like this 
```
 mediaIdList
                .asSequence()
                .map { mediaId ->
                    mediaStore.getSiteMediaWithId(site, mediaId)
                }
                .onEach { media ->
                    if (media == null) {
                        AppLog.w(T.MEDIA, "Cannot add null media to post")
                    } else {
                        editorTracker.trackAddMediaEvent(site, source, media)
                    }
                }
                .filterNotNull()
                .map { media ->
                    fluxCUtilsWrapper.mediaFileFromMediaModel(media)?.let { mediaFile -> Pair(media, mediaFile) }
                }
                .filterNotNull()
                .associateByTo(ArrayMap(), { it.first.urlToUse }, { it.second })
                .apply { editorMediaListener.appendMediaFiles(this) }
```

However, the result code is much longer and feels really over-engineered so it's probably not worth it. I'd just like to hear your thoughts about this issue in general, if you have any.

Reply:
I considered refactoring that, but decided it was not worth sacrificing the readability. I agree with your concerns about mutation in general, but I also believe context matters. As you said, the code block ^ is a little too verbose. I've taken a different approach in 6766b759380f22bc433e828c54e8432da8339c86 which can further be condensed, but I like the separation this provides. Let me know what your thoughts are!

```
val mediaMap = mediaIdList.asSequence().mapNotNull { mediaId ->
    mediaStore.getSiteMediaWithId(site, mediaId)
}.mapNotNull { media ->
    editorTracker.trackAddMediaEvent(site, source, media.isVideo)
    fluxCUtilsWrapper.mediaFileFromMediaModel(media)?.let { mediaFile ->
        Pair(media.urlToUse, mediaFile)
    }
}.toMap(ArrayMap())
(mediaIdList.size - mediaMap.size).takeIf { it > 0 }?.let { failedMediaCount ->
    AppLog.w(T.MEDIA, "Failed to add $failedMediaCount media to post")
}
editorMediaListener.appendMediaFiles(mediaMap)
```

Looking at it again, I guess the `trackAddMediaEvent` would have been better in a `forEach` on its own. There seems to be a lint issue, I'll make that change with the lint fix.